### PR TITLE
fixes #18826: containers do not restart after reboot…

### LINKED ISSUE #18826: containers do not restart after reboot when bound to virtual network interface
The restart policies `always` and `unless-stopped` do not work as expected for containers that have ports published on virtual interfaces if the host gets rebooted:

I have an Ubuntu 14.04 with a virtual interface added in `/etc/network/interfaces`:

```
auto eth0
iface eth0 inet static
        address 10.0.1.49
        netmask 255.255.0.0
        network 10.0.0.0
        broadcast 10.0.255.255
        gateway 10.0.0.1
        dns-nameservers 8.8.8.8 8.8.4.4

auto eth0:0
iface eth0:0 inet static
        address 10.0.81.0
        netmask 255.255.0.0
        broadcast 10.0.255.255
```

If I start a container with a port published on the "real" host IP, the container will restart after reboot:

```
$docker run -d -p 10.0.1.49:80:80 --restart unless-stopped nginx
```

But if I do the same using a virtual network interface IP, the container will not come up after a reboot:

```
$docker run -d -p 10.0.81.0:80:80 --restart unless-stopped nginx
65e2f7f16bfee320bd86b015897cdc62b16377fcaf08171439748e44136c8940
$sudo reboot
[...]
$ docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                       PORTS               NAMES
65e2f7f16bfe        nginx               "nginx -g 'daemon off"   6 minutes ago       Exited (128) 4 minutes ago                       backstabbing_keller

# in docker.log
ERRO[0000] Failed to start container 65e2f7f16bfee320bd86b015897cdc62b16377fcaf08171439748e44136c8940: failed to create endpoint backstabbing_keller on network bridge: Error starting userland proxy: listen tcp 10.0.81.0:80: bind: cannot assign requested address
```

This happens with the restart policies `always` and `unless-stopped`.
Note that the containers do come up after restarting the docker daemon but not after rebooting the host.

My environment:

```
$ docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64
```

```
$ docker info
Containers: 1
Images: 12
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 14
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-71-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 8
Total Memory: 3.858 GiB
Name: docker2
ID: 4AHB:OT2I:Z232:YHIQ:RSVQ:PX3O:57GF:N72P:A2AE:Z2JR:MBNQ:RHPX
Debug mode (server): true
 File Descriptors: 13
 Goroutines: 20
 System Time: 2015-12-21T15:46:07.8196672+01:00
 EventsListeners: 0
 Init SHA1: 4fc7c03a06c675f115b39958c13486d53df10a14
 Init Path: /usr/lib/docker/dockerinit
 Docker Root Dir: /var/lib/docker
Username: albers
Registry: https://index.docker.io/v1/
```

```
$ uname -a
Linux docker2 3.13.0-71-generic #114-Ubuntu SMP Tue Dec 1 02:34:22 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

### PR DESCRIPTION
**Pull request by Dmitry-k42**
…when bound to virtual network interface

Waiting for network is up when using systemd as described in [https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/).

In contrib/init/systemd/docker.service* string
`After=network.target docker.socket`
replaced to
`After=network-online.target docker.socket`
Probably fix for issue #18826

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks!

ping @andrewhsu PTAL

File: contrib/init/systemd/docker.service [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [Unit]
 Description=Docker Application Container Engine
 Documentation=https://docs.docker.com
-After=network.target docker.socket firewalld.service
+After=network-online.target docker.socket firewalld.service
```
Comment (thaJeztah) [inline contrib/init/systemd/docker.service]: Following the discussion in https://github.com/docker/docker/issues/29876#issue-198677228, @JamieGarside suggests to also add;

```
Wants=network-online.target
```

**Comment by thaJeztah** [comment]
ping @mavenugo @aaronlehmann PTAL as well

File: contrib/init/systemd/docker.service (hunk 1 above)
Comment (mentalMedley) [inline contrib/init/systemd/docker.service]: I tested this modification on my machine. It works as well as before. All containers autostarts correctly. Also I found [this about Wants](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Wants=):
> This is the recommended way to hook start-up of one unit to the start-up of another unit.

So Wants-option also must be added to Unit-file.

Comment (thaJeztah) [inline contrib/init/systemd/docker.service]: Can you update the PR with those changes? While you're doing that, can you squash your commits, so that there's a single commit in this PR? (some information on how to do that can be found in the [contributors guide](https://docs.docker.com/opensource/workflow/work-issue/#/pull-and-rebase-frequently))

Comment (mentalMedley) [inline contrib/init/systemd/docker.service]: I don't understand how to do git rebase. I have read contributors guide and @GordonTheTurtle 's advises and a lot of articles in Internet about it. I catch error "fatal: Needed a single revision
invalid upstream" and don't know what to do with this. It's too hard for me. Sorry :-( 
Please let @JamieGarside make his own pull request. I closed this PR.

**Comment by thaJeztah** [comment]
carrying this in https://github.com/docker/docker/issues/30210
